File: 3488-closest-equal-element-queries/solution.cpp
#include <vector>
#include <unordered_map>
#include <algorithm>

using namespace std;

class Solution {
public:
    vector<int> closestEqualElement(vector<int>& nums, vector<int>& queries) {
        int n = nums.size();
        // Map each value to a list of indices where it appears
        unordered_map<int, vector<int>> groups;
        for (int i = 0; i < n; ++i) {
            groups[nums[i]].push_back(i);
        }

        // Pre-calculate minimum distance for every index
        vector<int> min_dists(n, -1);
        for (auto& [val, pos] : groups) {
            int k = pos.size();
            if (k < 2) continue; // Unique element has no equal counterpart

            // The circular distance between the first and last occurrence
            int wrap_dist = n - (pos[k - 1] - pos[0]);

            for (int i = 0; i < k; ++i) {
                int dist_prev, dist_next;

                // Distance to the neighbor on the left
                if (i == 0) dist_prev = wrap_dist;
                else dist_prev = pos[i] - pos[i - 1];

                // Distance to the neighbor on the right
                if (i == k - 1) dist_next = wrap_dist;
                else dist_next = pos[i + 1] - pos[i];

                min_dists[pos[i]] = min(dist_prev, dist_next);
            }
        }

        // Answer queries from pre-calculated data
        vector<int> results;
        results.reserve(queries.size());
        for (int q : queries) {
            results.push_back(min_dists[q]);
        }

        return results;
    }
};

File: 3488-closest-equal-element-queries/solution.py
from collections import defaultdict

class Solution:
    def closestEqualElement(self, nums: list[int], queries: list[int]) -> list[int]:
        n = len(nums)
        # Group indices by their values
        groups = defaultdict(list)
        for idx, val in enumerate(nums):
            groups[val].append(idx)
        
        # Pre-calculate the result for every index in the array
        # Initialize with -1 for values that appear only once
        ans_at_index = [-1] * n
        
        for val, positions in groups.items():
            k = len(positions)
            if k < 2:
                continue
            
            # Distance between the last and first index across the circular boundary
            wrap_dist = n - (positions[-1] - positions[0])
            
            for i in range(k):
                # Calculate distance to previous occurrence
                if i == 0:
                    dist_prev = wrap_dist
                else:
                    dist_prev = positions[i] - positions[i-1]
                
                # Calculate distance to next occurrence
                if i == k - 1:
                    dist_next = wrap_dist
                else:
                    dist_next = positions[i+1] - positions[i]
                
                # Store the minimum of the two circular neighbors
                ans_at_index[positions[i]] = min(dist_prev, dist_next)
        
        # Map queries to the pre-calculated answers
        return [ans_at_index[q] for q in queries]